File: javascript/80.remove-duplicates-from-sorted-array-ii.js
/*
 * @lc app=leetcode id=80 lang=javascript
 *
 * [80] Remove Duplicates from Sorted Array II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function (nums) {
  if (nums.length <= 2) return nums.length;
  let current = nums[0],
    s = 1,
    count = 1;
  for (let f = 1; f < nums.length; f++) {
    if (nums[f] !== current) {
      current = nums[f];
      [nums[s], nums[f]] = [nums[f], nums[s]];
      s++;
      count = 1;
    } else if (count < 2) {
      [nums[s], nums[f]] = [nums[f], nums[s]];
      s++;
      count++;
    }
  }
  return s;
};
// @lc code=end

File: rust/80.remove-duplicates-from-sorted-array-ii.rs
/*
 * @lc app=leetcode id=80 lang=rust
 *
 * [80] Remove Duplicates from Sorted Array II
 */

// @lc code=start
impl Solution {
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        if nums.len() <= 2 {
            return nums.len() as i32;
        }
        let mut current = nums[0];
        let mut count = 1;
        let mut s = 1;
        for f in 1..nums.len() {
            if nums[f] != current {
                current = nums[f];
                nums.swap(s, f);
                count = 1;
                s += 1;
            } else if count < 2 {
                nums.swap(s, f);
                count += 1;
                s += 1;
            }
        }
        s as i32
    }
}
// @lc code=end
